File: CSharp/Engine/Runtime/Asset/Bitmap.cs
using System;
using System.IO;


namespace Engine
{
    public enum BitmapFormat
    {
        B8G8R8,
        B8G8R8A8,
    }


    /// <summary>
    /// Encode/Decode uncompressed 24-bit(B8G8R8) or 32-bit(B8G8R8A8) *.bmp file
    /// </summary>
    public class Bitmap
    {
        public const int FileHeaderSize		= 14;
        public const int InfoHeaderSize		= 40;
        public const int DefaultBitsOffset	= FileHeaderSize + InfoHeaderSize;


        public int Width { get; private set; }
        public int Height { get; private set; }
        public byte[] Buffer { get; private set; }
        public BitmapFormat Format { get; private set; }
        public int ChannelCount { get; private set; }


        public Bitmap(int width, int height, BitmapFormat format)
        {
            int channelCount = format == BitmapFormat.B8G8R8 ? 3 : 4;

            Width			= width;
            Height			= height;
            Buffer			= new byte[ channelCount * width * height ];
            Format			= format;
            ChannelCount	= channelCount;
        }


        /// <summary>
        /// The pixel's high 8-bit is unused if Format == BitmapFormat.B8G8R8
        /// </summary>
        public void SetPixel(int x, int y, uint pixel)
        {
            for ( int index = 0; index < ChannelCount; index++ )
                Buffer[ ChannelCount * ( Width * y + x ) + index ] = (byte)( ( pixel >> ( 8 * index ) ) & 0xFF );
        }

        public uint GetPixel(int x, int y)
        {
            uint ret = 0;
            for ( int index = 0; index < ChannelCount; index++ )
                ret |= (uint)( Buffer[ ChannelCount * ( Width * y + x ) + index ] << ( 8 * index ) );

            return ret;
        }

        public void FlipHorizontal()
        {
            var Fn = (ref byte a, ref byte b) =>
            {
                byte temporary	= a;
                a				= b;
                b				= temporary;
            };

            int halfWidth = Width / 2;
            for ( int y = 0; y < Height; y++ )
                for ( int x = 0; x < halfWidth; x++ )
                    for ( int index = 0; index < ChannelCount; index++ )
                        Fn(
                            ref Buffer[ ChannelCount * ( Width * y + x ) + index ],
                            ref Buffer[ ChannelCount * ( Width * y + ( Width - 1 - x ) ) + index ]
                            );
        }

        public void FlipVertical()
        {
            var Fn = (ref byte a, ref byte b) =>
            {
                byte temporary	= a;
                a				= b;
                b				= temporary;
            };

            int halfHeight = Height / 2;
            for ( int y = 0; y < halfHeight; y++ )
                for ( int x = 0; x < Width; x++ )
                    for ( int i = 0; i < ChannelCount; i++ )
                        Fn(
                            ref Buffer[ ChannelCount * ( Width * y + x ) + i ],
                            ref Buffer[ ChannelCount * ( Width * ( Height - 1 - y ) + x ) + i ]
                            );
        }


        /// <summary>
        /// Test if buffer is a valid *.bmp file representation
        /// </summary>
        public static bool IsValid(byte[] buffer)
        {
            if ( buffer.Length < DefaultBitsOffset )
                return false;

            if ( buffer[0] != 'B' || buffer[1] != 'M' )
                return false;

            int bitCount = BitConverter.ToInt32(buffer, 28);
            if ( bitCount != 24 && bitCount != 32 )
                return false;

            return true;
        }

        public static byte[] Encode(Bitmap instance)
        {
            int imageSize = instance.ChannelCount * instance.Width * instance.Height;
            int byteCount = DefaultBitsOffset + imageSize;

            using( var ms = new MemoryStream(byteCount) )
            using ( var writer = new BinaryWriter(ms) )
            {
                // set Bitmap FileHeader
                writer.Write( "BM".ToCharArray() );
                writer.Write(byteCount);
                writer.Write(0);
                writer.Write(DefaultBitsOffset);

                // set Bitmap InfoHeader
                writer.Write(InfoHeaderSize);
                writer.Write( instance.Width );
                writer.Write( instance.Height );
                writer.Write( (short)1 );
                writer.Write( (short)( 8 * instance.ChannelCount ) );
                writer.Write(0);
                writer.Write(imageSize);
                writer.Write(0);
                writer.Write(0);
                writer.Write(0);
                writer.Write(0);

                // set image data
                writer.Write( instance.Buffer );

                return ms.ToArray();
            }
        }

        public static Bitmap Decode(byte[] buffer)
        {
            int offset = BitConverter.ToInt32(buffer, 10);

            int width			= BitConverter.ToInt32(buffer, 18);
            int height			= BitConverter.ToInt32(buffer, 22);
            int bitCount		= BitConverter.ToInt32(buffer, 28);
            BitmapFormat format	= bitCount == 24 ? BitmapFormat.B8G8R8 : BitmapFormat.B8G8R8A8;

            var ret = new Bitmap(width, height, format);

            for ( int index = 0; index < ret.Buffer.Length; index++ )
                ret.Buffer[index] = buffer[ offset + index ];

            return ret;
        }
    }
}

File: C/Engine/Source/Runtime/Asset/Bitmap.c
#include "Bitmap.h"



static const SizeType BitmapFileHeaderSize		= sizeof(BitmapFileHeader);
static const SizeType BitmapInfoHeaderSize		= sizeof(BitmapInfoHeader);
static const SizeType BitmapDefaultBitsOffset	= sizeof(BitmapFileHeader) + sizeof(BitmapInfoHeader);


static Bitmap Construct(SizeType width, SizeType height, BitmapFormat format)
{
	SizeType channelCount = format == BitmapFormat_B8G8R8 ? 3 : 4;

	Bitmap ret;
	ret.buffer			= FAllocator.Allocate( channelCount * width * height );
	ret.width			= width;
	ret.height			= height;
	ret.format			= format;
	ret.channelCount	= channelCount;

	return ret;
}

static Void Destruct(Bitmap* instance)
{
	if ( !instance )
		return;

	FAllocator.Release( instance->buffer );
}

static Void SetPixel(Bitmap* instance, SizeType x, SizeType y, UInt32 pixel)
{
	FMemory.Copy(
		&pixel,
		CAST( Byte*, instance->buffer ) + instance->channelCount * ( instance->width * y + x ),
		instance->channelCount
	);
}

static UInt32 GetPixel(const Bitmap* instance, SizeType x, SizeType y)
{
	UInt32 ret;
	FMemory.Copy(
		CAST( Byte*, instance->buffer ) + instance->channelCount * ( instance->width * y + x ),
		&ret,
		instance->channelCount
	);

	return ret;
}

static Void FlipHorizontal(Bitmap* instance)
{
	SizeType width		= instance->width;
	SizeType height		= instance->height;
	SizeType halfWidth	= width / 2;

	for ( SizeType y = 0; y < height; y++ )
		for ( SizeType x = 0; x < halfWidth; x++ )
			for ( SizeType index = 0; index < instance->channelCount; index++ )
				SWAP(
					Byte,
					CAST(Byte*, instance->buffer)[ instance->channelCount * ( width * y + x ) + index ],
					CAST(Byte*, instance->buffer)[ instance->channelCount * ( width * y + ( width - 1 - x ) ) + index ]
					);
}

static Void FlipVertical(Bitmap* instance)
{
	SizeType width		= instance->width;
	SizeType height		= instance->height;
	SizeType halfHeight	= height / 2;

	for ( SizeType y = 0; y < halfHeight; y++ )
		for ( SizeType x = 0; x < width; x++ )
			for ( SizeType index = 0; index < instance->channelCount; index++ )
				SWAP(
					Byte,
					CAST(Byte*, instance->buffer)[ instance->channelCount * ( width * y + x ) + index ],
					CAST(Byte*, instance->buffer)[ instance->channelCount * ( width * ( height - 1 - y ) + x ) + index ]
				);
}

static Bool IsValid(const Void* buffer, SizeType byteCount)
{
	if ( byteCount < BitmapDefaultBitsOffset )
		return False;

	if ( !FMemory.Equal( buffer, "BM", 2 ) )
		return False;

	BitmapInfoHeader* infoHeader = REINTERPRET_CAST(const BitmapInfoHeader*, buffer, BitmapFileHeaderSize);
	if (
		infoHeader->bitCount != 24 &&
		infoHeader->bitCount != 32
		)
		return False;

	return True;
}

static Bitmap Decode(const Void* buffer, SizeType byteCount)
{
	BitmapInfoHeader* infoHeader = REINTERPRET_CAST(const BitmapInfoHeader*, buffer, BitmapFileHeaderSize);

	BitmapFormat format		= infoHeader->bitCount == 24 ? BitmapFormat_B8G8R8 : BitmapFormat_B8G8R8A8;
	SizeType channelCount	= infoHeader->bitCount / 8;

	Void* image = FAllocator.Allocate( infoHeader->imageSize );
	FMemory.Copy(
		CAST(const Byte*, buffer) + CAST(const BitmapFileHeader*, buffer)->offset,
		image,
		infoHeader->imageSize
	);

	Bitmap ret = { image, infoHeader->width, infoHeader->height, format, channelCount };
	return ret;
}

static HeapMemory Encode(const Bitmap* instance)
{
	SizeType imageSize = instance->channelCount * instance->width * instance->height;
	SizeType byteCount = BitmapDefaultBitsOffset + imageSize;

	HeapMemory ret = FHeapMemory.Construct(byteCount);
	{
		Void* buffer = ret.buffer;
		FMemory.Clear(buffer, BitmapDefaultBitsOffset);
		FMemory.Copy(
			instance->buffer,
			CAST(Byte*, buffer) + BitmapDefaultBitsOffset,
			imageSize
		);

		// set BitmapFileHeader
		BitmapFileHeader* fileHeader	= CAST(BitmapFileHeader*, buffer);
		fileHeader->magic[0]			= 'B';
		fileHeader->magic[1]			= 'M';
		fileHeader->size				= byteCount;
		fileHeader->offset				= BitmapDefaultBitsOffset;

		// set BitmapInfoHeader
		BitmapInfoHeader* infoHeader	= REINTERPRET_CAST(BitmapInfoHeader*, buffer, BitmapFileHeaderSize);
		infoHeader->size				= BitmapInfoHeaderSize;
		infoHeader->width				= instance->width;
		infoHeader->height				= instance->height;
		infoHeader->planes				= 1;
		infoHeader->bitCount			= 8 * instance->channelCount;
		infoHeader->imageSize			= imageSize;
	}

	return ret;
}


struct FBitmap FBitmap = 
{
	Construct,
	Destruct,
	SetPixel,
	GetPixel,
	FlipHorizontal,
	FlipVertical,
	IsValid,
	Decode,
	Encode,
};
